### PR TITLE
Gh5757 Invoice Candidate Status Endpoint

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/5757

### CODE REVIEW

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		return JsonGetInvoiceCandidatesStatusResponse.ok(result);
+	}
+
+	private List<JsonInvoiceCandidateStatus> createStatus(final List<InvoiceCandidateId> invoiceCandidateIds)
+	{
+		final List<JsonInvoiceCandidateStatus> invoiceCandidatesStatus = new ArrayList<>();
+
+		for (final InvoiceCandidateId invoiceCandidateId : invoiceCandidateIds)
+		{
+			final I_C_Invoice_Candidate invoiceCandidateRecord = invoiceCandDAO.getById(invoiceCandidateId);
```
Comment (teosarca) [inline de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java]: avoid N+1

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		}
+
+		return workPackageRecords
+				.stream()
+				.map(workPackageRecord -> createJsonForWorkPackage(workPackageRecord))
+				.collect(ImmutableList.toImmutableList());
+
+	}
+
+	private List<I_C_Queue_WorkPackage> getWorkProcessesForInvoiceCandidate(final I_C_Invoice_Candidate invoiceCandidateRecord)
```
Comment (teosarca) [inline de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java]: this shall go into IQueueDAO

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		if (invoiceLinesForInvoiceCandidate.isEmpty())
+		{
+
+			return ImmutableList.of();
+		}
+
+		return invoiceLinesForInvoiceCandidate
+				.stream()
+				.map(I_C_InvoiceLine::getC_Invoice_ID)
+				.map(invoiceId -> invoiceDAO.getByIdInTrx(InvoiceId.ofRepoId(invoiceId)))
```
Comment (teosarca) [inline de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java]: avoid N+1

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java (hunk 1 above)
Comment (metas-rc) [reply de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java]: Solved

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java (hunk 2 above)
Comment (metas-rc) [reply de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java]: solved

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java (hunk 3 above)
Comment (metas-rc) [reply de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/InvoiceCandidateStatusRepository.java]: solved

File: de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java [hunk 4]
```diff
@@ -130,14 +132,21 @@
...
 		Check.assumeNotNull(packageProcessorClass, "packageProcessorClass not null");
 
 		final String packageProcessorClassname = packageProcessorClass.getCanonicalName();
 		return retrievePackageProcessorDefByClassname(ctx, packageProcessorClassname);
 	}
 
+	@Override
+	public int retrievePackageProcessorIdByClass(@NonNull final Class<? extends IWorkpackageProcessor> packageProcessorClass)
+	{
+		Check.assumeNotNull(packageProcessorClass, "packageProcessorClass not null");
```
Comment (metas-ts) [inline de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java]:
```suggestion
```
this is what the `@NonNull` annotatin does for you already (so, if `packageProcessorClass` *was* null, the `assumeNotNull`) would not be reached

File: de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java [hunk 5]
```diff
@@ -370,4 +379,25 @@ public boolean isWorkpackageProcessorEnabled(final Class<? extends IWorkpackageP
...
+
+		final int workpackageProcessorId = retrievePackageProcessorIdByClass(packageProcessorClass);
+
+		return queryBL.createQueryBuilder(I_C_Queue_Element.class)
+				.addEqualsFilter(I_C_Queue_Element.COLUMNNAME_AD_Table_ID, recordRef.getAD_Table_ID())
+				.addEqualsFilter(I_C_Queue_Element.COLUMNNAME_Record_ID, recordRef.getRecord_ID())
+				.andCollect(I_C_Queue_Element.COLUMN_C_Queue_WorkPackage_ID)
+				.addNotEqualsFilter(I_C_Queue_WorkPackage.COLUMNNAME_Processed, true)
+				.andCollect(I_C_Queue_WorkPackage.COLUMN_C_Queue_Block_ID)
+				.addEqualsFilter(I_C_Queue_Block.COLUMNNAME_C_Queue_PackageProcessor_ID, workpackageProcessorId)
```
Comment (metas-ts) [inline de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java]: here you go from `C_Queue_WorkPackage` up to `C_Queue_Block` and then 
back down to `C_Queue_WorkPackage`.
Now you got *all* the WPs that have this C_Queue_Block_ID, even if not all of them have an element with the given `recordRef`.

We *often* have just one WP per block, but that's not the case in general.
So i think you need to modify this a bit.
IDK by heart to do this...maybe you can work with a subquery..ping me if you run into problems...

File: de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonGetInvoiceCandidatesStatusRequest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+@Value
+public class JsonGetInvoiceCandidatesStatusRequest
+{
+	@ApiModelProperty(position = 10, required = true, //
+			value = "Specifies the invoice candidates to be invoiced.")
```
Comment (metas-ts) [inline de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonGetInvoiceCandidatesStatusRequest.java]:
```suggestion
			value = "Specifies the invoice candidates to return the invoicing status of.")
```

File: de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java (hunk 4 above)
Comment (metas-rc) [reply de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java]: fixed.

File: de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonGetInvoiceCandidatesStatusRequest.java (hunk 6 above)
Comment (metas-rc) [reply de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonGetInvoiceCandidatesStatusRequest.java]: done

File: de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java (hunk 5 above)
Comment (metas-rc) [reply de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java]: done

File: de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonGetInvoiceCandidatesStatusRequest.java (hunk 6 above)
Comment (metas-rc) [reply de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonGetInvoiceCandidatesStatusRequest.java]: now the generated sql would look like 

```
SELECT ad_client_id,ad_issue_id,ad_org_id,ad_pinstance_id,ad_role_id,ad_user_id,ad_user_incharge_id,batchenqueuedcount,c_async_batch_id,c_queue_block_id,
( 
       SELECT c_queue_block.c_queue_packageprocessor_id 
       FROM   c_queue_block 
       WHERE  c_queue_block.c_queue_block_id=c_queue_workpackage.c_queue_block_id) AS c_queue_packageprocessor_id,c_queue_workpackage_id,created,createdby,errormsg,isactive,iserror,iserroracknowledged,isreadyforprocessing,lastdurationmillis,lastendtime, laststarttime,priority,processed,skipped_count,skipped_first_time,skipped_last_reason,skippedat,skiptimeoutmillis,updated,updatedby FROM c_queue_workpackage WHERE (
  ( 
    EXISTS 
    ( 
           SELECT 1 
           FROM   c_queue_element 
           WHERE  ( 
                         c_queue_element.c_queue_workpackage_id = c_queue_workpackage.c_queue_workpackage_id
                  AND    c_queue_workpackage_id IS NOT NULL 
                  AND    (( 
                                       ad_table_id = 540270) 
                         AND    ( 
                                       record_id = 1000033)))) 
  ) 
  AND 
  ( 
    ( 
      processed <> 'Y' 
      OR 
      processed IS NULL 
    ) 
  ) 
  AND 
  ( 
    EXISTS 
    ( 
           SELECT 1 
           FROM   c_queue_block 
           WHERE  ( 
                         c_queue_block.c_queue_block_id = c_queue_workpackage.c_queue_block_id 
                  AND    c_queue_block_id IS NOT NULL 
                  AND    (( 
                                       c_queue_packageprocessor_id = 540000)))) 
  ) 
)
```

File: de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java (hunk 5 above)
Comment (metas-rc) [reply de.metas.async/src/main/java/de/metas/async/api/impl/AbstractQueueDAO.java]: now the generated sql would look like 

```
SELECT ad_client_id,ad_issue_id,ad_org_id,ad_pinstance_id,ad_role_id,ad_user_id,ad_user_incharge_id,batchenqueuedcount,c_async_batch_id,c_queue_block_id,
( 
       SELECT c_queue_block.c_queue_packageprocessor_id 
       FROM   c_queue_block 
       WHERE  c_queue_block.c_queue_block_id=c_queue_workpackage.c_queue_block_id) AS c_queue_packageprocessor_id,c_queue_workpackage_id,created,createdby,errormsg,isactive,iserror,iserroracknowledged,isreadyforprocessing,lastdurationmillis,lastendtime, laststarttime,priority,processed,skipped_count,skipped_first_time,skipped_last_reason,skippedat,skiptimeoutmillis,updated,updatedby FROM c_queue_workpackage WHERE (
  ( 
    EXISTS 
    ( 
           SELECT 1 
           FROM   c_queue_element 
           WHERE  ( 
                         c_queue_element.c_queue_workpackage_id = c_queue_workpackage.c_queue_workpackage_id
                  AND    c_queue_workpackage_id IS NOT NULL 
                  AND    (( 
                                       ad_table_id = 540270) 
                         AND    ( 
                                       record_id = 1000033)))) 
  ) 
  AND 
  ( 
    ( 
      processed <> 'Y' 
      OR 
      processed IS NULL 
    ) 
  ) 
  AND 
  ( 
    EXISTS 
    ( 
           SELECT 1 
           FROM   c_queue_block 
           WHERE  ( 
                         c_queue_block.c_queue_block_id = c_queue_workpackage.c_queue_block_id 
                  AND    c_queue_block_id IS NOT NULL 
                  AND    (( 
                                       c_queue_packageprocessor_id = 540000)))) 
  ) 
)
```

**Pull request merged**